webapp/app.py

```diff
@@ -73,11 +73,28 @@ def thank_you():
 
 @open_id.after_login
 def after_login(resp):
-    flask.session["macaroon_discharge"] = resp.extensions["macaroon"].discharge
-    flask.session["openid"] = {
-        "identity_url": resp.identity_url,
-        "email": resp.email,
+    root = flask.session["macaroon_root"]
+    discharge = resp.extensions["macaroon"].discharge
+    authorization_header = get_authorization_header(root, discharge)
+    authorization_code = f"root={root} discharge={discharge}"
+    invitation_code = flask.session["invitation_code"]
+    data = {
+        "provider": "usso",
+        "authorization_code": authorization_code,
+        "invitation_code": invitation_code,
     }
+
+    response = requests.post(
+        url=f"{ANBOXCLOUD_API_BASE}1.0/login",
+        headers=authorization_header,
+        json=data,
+    )
+    body = response.json()
```

Comment:
Can you create a function that will make the API call and returns the `body.json()`? This will make it easier to read.

Reply:
Personally I disagree.

When extract code into a function, you're trading off two things - you're making the code shorter and quicker to read in the original location, but you're also obfuscating what's actually happening.

(Of course, the *main* reason for creating a function is because you want to re-use the code elsewhere - as long as it's more than a couple of lines, this is always a good idea, but I don't think that's what we're talking about here)

So the question basically comes down to - might this code actually be helpful to the person reading the function? Or is it just distracting to them, and they'd rather see a function name where they don't know what's actually happening in the function?

Now this might depend a little bit on your level of experience with Python or how quickly you're able to skim code, whether you're familiar with requests etc.

But personally, I think these 6 lines, or 2 code blocks, is actually pretty useful to see directly. I can tell that requests is the library, what the API endpoint being called is, that it returns JSON, all at a glance. And given that not a whole lot is happening in this function, that API endpoint being called seems like just about the *most* interesting detail of what's going on here - I'd very likely be looking for that piece of information if I was reading this function.

And personally, if I wasn't looking for this piece, skipping over this block wouldn't be particularly difficult, I don't think it would slow down my reading of the rest of the code by much.

Reply:
However, you should probably add `response.raise_for_status()`

Reply:
Yeah. I also had it in a function before, but for me both cases are ok. But the fact is, we are not reusing the same function everywhere, there is one POST with payload, another GET request without payload and they even have different headers, so each request is quite different than the other in terms of parameters passed. 

If I had to create a generic function to reuse it, my idea would be to use `requests("METHOD")`, which is actually the lowest layer in the library, so I don't see much sense as of now to do that. Maybe after we have a bunch of endpoints I could start doing that.

Reply:
@carkod although I think you could probably help a little bit with @tbille's concern of readability if this function by separating our logic parts with blank lines (and maybe also consider some comments). E.g.:

``` python3
@open_id.after_login
def after_login(resp):
    root = flask.session.pop("macaroon_root")
    discharge = resp.extensions["macaroon"].discharge

    # Get authorization_token from API
    data = {
        "provider": "usso",
        "authorization_code": f"root={root} discharge={discharge}",
        "invitation_code": flask.session.pop("invitation_code"),
    }
    response = requests.post(url=f"{ANBOXCLOUD_API_BASE}1.0/login", json=data,)
    response.raise_for_status()

    flask.session["authentication_token"] = response.json()["metadata"]["token"]

    return flask.redirect(open_id.get_next_url())
```

You'll also notice that I've changed a little the way you get the items from the session. I don't think you should do `flaks.session.clear()`, as we can't be sure other parts of the app might not be using the session to store other data. All we know is that we no longer need the 'macaroon_root' and 'invitation_code' parts which are part of this login flow. The neatest way to communicate that, to me, is to use the `.pop()` method, which will remove and return the item from the session.

I also removed a couple of lines by eliminating a couple of single-use variables.

Reply:
Also here, as mentioned below, there is a need to catch the 401 basically that is the whole point of using invitation codes, to catch users that are not allowed access. otherwise, they say that anyone could create a SSO account and view the demo, which is not desired

Reply:
Good point. In which case you want:

``` python3
try:
    response.raise_for_status()
except HTTPError as http_error:
    if http_error.response.status_code == 401:
        flask.abort(401, http_error.response.json()["error"])

    raise http_error
```

Like in the other case.